libwallet/src/slatepack/address.rs

```diff
@@ -0,0 +1,220 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use bech32::{self, FromBase32, ToBase32};
+/// Slatepack Address definition
+use ed25519_dalek::PublicKey as edDalekPublicKey;
+use ed25519_dalek::SecretKey as edDalekSecretKey;
+use rand::{thread_rng, Rng};
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use x25519_dalek::PublicKey as xDalekPublicKey;
+
+use crate::grin_core::ser::{self, Readable, Reader, Writeable, Writer};
+use crate::util::OnionV3Address;
+use crate::{Error, ErrorKind};
+
+use std::convert::TryFrom;
+use std::fmt::{self, Display};
+
+/// Definition of a Slatepack address
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct SlatepackAddress {
+	/// Human-readable prefix
+	pub hrp: String,
+	/// ed25519 Public key, to be bech32 encoded,
+	/// interpreted as tor address or converted
+	/// to an X25519 public key for encrypting
+	/// slatepacks
+	pub pub_key: edDalekPublicKey,
+}
+
+impl SlatepackAddress {
+	/// new with default hrp
+	pub fn new(pub_key: &edDalekPublicKey) -> Self {
+		Self {
+			hrp: String::from("slatepack"),
+			pub_key: pub_key.clone(),
+		}
+	}
+
+	/// new with a random key
+	pub fn random() -> Self {
+		let bytes: [u8; 32] = thread_rng().gen();
+		let pub_key = edDalekPublicKey::from(&edDalekSecretKey::from_bytes(&bytes).unwrap());
+		SlatepackAddress::new(&pub_key)
+	}
+
+	/// calculate encoded length
+	pub fn encoded_len(&self) -> Result<usize, Error> {
+		let encoded = String::try_from(self)?;
+		// add length byte
+		Ok(encoded.as_bytes().len() + 1)
+	}
+}
+
+impl Display for SlatepackAddress {
+	fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+		formatter.write_str(&String::try_from(self).unwrap())
+	}
+}
+
+impl TryFrom<&str> for SlatepackAddress {
+	type Error = Error;
+	fn try_from(encoded: &str) -> Result<Self, Self::Error> {
+		let (hrp, data) = bech32::decode(&encoded)?;
+		let bytes = Vec::<u8>::from_base32(&data)?;
+		let pub_key = match edDalekPublicKey::from_bytes(&bytes) {
+			Ok(k) => k,
+			Err(e) => {
+				return Err(ErrorKind::ED25519Key(format!("{}", e)).into());
+			}
+		};
+		Ok(SlatepackAddress { hrp, pub_key })
+	}
+}
+
+impl TryFrom<&SlatepackAddress> for String {
+	type Error = Error;
+	fn try_from(addr: &SlatepackAddress) -> Result<Self, Self::Error> {
+		let encoded = bech32::encode(&addr.hrp, addr.pub_key.to_bytes().to_base32())?;
+		Ok(encoded.to_string())
+	}
+}
+
+impl From<&SlatepackAddress> for OnionV3Address {
+	fn from(addr: &SlatepackAddress) -> Self {
+		OnionV3Address::from_bytes(addr.pub_key.to_bytes())
+	}
+}
+
+impl TryFrom<&SlatepackAddress> for xDalekPublicKey {
+	type Error = Error;
```

Comment:
Bonus point here for telling me why this conversion doesn't seem to work.

Reply:
It does work, the ed25519 key just needs to be SHA512ed before using it as a decrypting key for x25519